opentelemetry-sdk/src/opentelemetry/sdk/util.py

```diff
@@ -26,6 +28,31 @@
     from collections import Sequence
 
 
+def set_timeout_signal_handler():
+    "Signal timeout setter."
+
+    def signal_handler_function(signum, frame):
+        raise TimeoutError
+
+    return signal.signal(signal.SIGALRM, signal_handler_function)
+
+
+@contextmanager
+def timeout_in_seconds(seconds=None):
```

Comment:
The idea of having a general purpose timeout mechanism is great, but keep in mind that the specification requires `export` and `shutdown` to not block indefinitely. This approach (using a `@contextmanager` to make a context object that is used when `export` and `shutdown` are called) puts the responsibility of being specification compliant on the code that does the _calling_ of `export` and `shutdown` instead of putting the responsibility on these 2 methods themselves not to block indefinitely.

Please consider making a decorator that is then applied directly to the functions `export` and `shutdown`. In this way, they can be called directly and the caller won't have to worry about making them not block indefinitely, something like this:

```python
def timeout(function):

    def inner(*args, timeout=90, **kwargs):

        print(timeout)

        function(*args, **kwargs)

    return inner


@timeout
def export(first, second):
    print(first)
    print(second)


export(1, 2, timeout=9)
```
```
9
1
2
```

Reply:
Exporters might have a better way to implement timeouts, e.g. in the zipkin exporter we'd pass a `timeout` arg to [`requests.post`](https://github.com/c24t/opentelemetry-python/blob/6aa69ae4bc07d07bf33652435c5fa63b65eb0190/ext/opentelemetry-ext-zipkin/src/opentelemetry/ext/zipkin/__init__.py#L86-L88) which sets a timeout on the underlying socket.

I think a general purpose timeout in processor is fine, but exporters should implement timeout logic themselves: https://github.com/open-telemetry/opentelemetry-python/issues/346#issuecomment-579998140.